File: 剑指offer/LinkedList/025-复杂链表的复制/code.py
class RandomListNode:
    def __init__(self, x)
    self.label = x
    self.next = None
    self.random = None


def Clone(pHead):
    if not pHead:
        return None
    
    p = pHead
    new_h = RandomListNode(p.label)
    pre_p = new_h
    random_map = {pHead:new_h}
    p = p.next
    while p:
        new_p = RandomListNode(p.label)
        random_map[p] = new_p
        pre_p.next = new_p
        pre_p =pre_p.next
        p = p.next
    p = pHead
    new_p = new_h
    while p:
        random_p = p.random
        if random_p:
            new_p.random = random_map[random_p]
        p = p.next
        new_p = new_p.next

File: 剑指offer/LinkedList/025-复杂链表的复制/code.cpp
class Solution{
    public:
    void CloneNodes(RandomListNode* pHead)
    {
        while(pNode != NULL)
        {
            RandomListNode * pCloned = new RandomListNode(0);
            pCloned->label = pNode->label;
            pCloned->next = pNode->next;
            pCloned->random = NULL;

            pNode->next = pCloned;
            pNode = pCloned->next;
        }
    }

    void ConnectSibingNodes(RandomListNode* pHead)
    {
        RandomListNode* Node = pHead;
        while(pNode != NULL)
        {
            RandomListNode * pCloned = pNode->next;
            if(pNode->random != NULL)
            {
                pCloned->random = pNode->random->next;
            }
            pNode = pCloned->next;
        }
    }
    RandomListNode * ReconnectNodes(RandomListNode* pHead)
    {
        RandomListNode* pNode = pHead;
        RandomListNode* pClonedHead = NULL;
        RandomListNode* pClonedNode = NULL;

        if(pNode != NULL){
            pClonedHead = pClonedNode = pNode->next;
            pNode->next = pClonedNode->next;
            pNode = pNode->next;
        }

        while(pNode != NULL)
        {
            pClonedNode->next = pNode->next;
            pClonedNode = pClonedNode->next;
            pNode->next = pClonedNode->next;
            pNode = pNode->next;
        }
        return pClonedHead;
    }

    RandomListNode * Clone(RandomListNode* pHead)
    {
        CloneNodes(pHead);
        ConnectSilingNodes(pHead);
        return ReconnectNodes(pHead);
    }
};